liquidity/fees.go

```diff
@@ -200,3 +204,127 @@ func (f *FeeCategoryLimit) loopOutFees(amount btcutil.Amount,
 
 	return prepayMaxFee, routeMaxFee, f.MaximumMinerFee
 }
+
+// FeePortion is a fee limitation which limits fees to a set portion of
+// the swap amount.
+type FeePortion struct {
+	// PartsPerMillion is the total portion of the swap amount that the
+	// swap may consume.
+	PartsPerMillion int
+}
+
+// NewFeePortion creates a fee limit based on a flat percentage of swap amount.
+func NewFeePortion(ppm int) *FeePortion {
+	return &FeePortion{
+		PartsPerMillion: ppm,
+	}
+}
+
+// String returns a string representation of the fee limit.
+func (f *FeePortion) String() string {
+	return fmt.Sprintf("parts per million: %v", f.PartsPerMillion)
+}
+
+// validate returns an error if the values provided are invalid.
+func (f *FeePortion) validate() error {
+	if f.PartsPerMillion <= 0 {
+		return ErrInvalidPPM
+	}
+
+	return nil
+}
+
+// loopOutLimits checks whether the quote provided is within our fee
+// limits for the swap amount.
+func (f *FeePortion) loopOutLimits(swapAmt btcutil.Amount,
+	quote *loop.LoopOutQuote) error {
+
+	// First, check whether any of the individual fee categories provided
+	// by the server are more than our total limit. We do this so that we
+	// can provide more specific reasons for not executing swaps.
+	feeLimit := ppmToSat(swapAmt, f.PartsPerMillion)
+
+	if quote.MinerFee > feeLimit {
+		log.Debugf("miner fee: %v greater than fee limit: %v, at "+
+			"%v ppm", quote.MinerFee, feeLimit, f.PartsPerMillion)
+
+		return newReasonError(ReasonMinerFee)
+	}
+
+	if quote.SwapFee > feeLimit {
+		log.Debugf("swap fee: %v greater than fee limit: %v, at "+
+			"%v ppm", quote.SwapFee, feeLimit, f.PartsPerMillion)
+
+		return newReasonError(ReasonSwapFee)
+	}
+
+	if quote.PrepayAmount > feeLimit {
+		log.Debugf("prepay amount: %v greater than fee limit: %v, at "+
+			"%v ppm", quote.PrepayAmount, feeLimit, f.PartsPerMillion)
+
+		return newReasonError(ReasonPrepay)
+	}
+
+	// If our miner and swap fee equal our limit, we will have nothing left
+	// for off-chain fees, so we fail out early.
+	if quote.MinerFee+quote.SwapFee >= feeLimit {
```

Comment:
Great approach of doing the checks incrementally, this should make it very clear to the user why a swap went through or not :clap: